src/typechecker/FStar.TypeChecker.TcTerm.fst

```diff
@@ -3249,7 +3249,7 @@ and tc_pat env (pat_t:typ) (p0:pat) :
         check_nested_pattern
             (Env.clear_expected_typ env |> fst)
             (PatternUtils.elaborate_pat env p0)
-            pat_t
+            (expected_pat_typ env p0.p pat_t)
```

Comment:
This looks interesting too ...

I guess it is doing something similar to computing the unrefined type of the scrutinee? 

I wonder if it could be simplified to what you've done in the patch to Rel, i.e., normalize_refinement |> unrefine.